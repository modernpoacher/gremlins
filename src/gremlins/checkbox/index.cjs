require('@babel/register')({
  ignore: [
    /node_modules\/(?!@modernpoacher\/cogs|@modernpoacher\/sprockets|@modernpoacher\/gremlins)/
  ]
})

const debug = require('debug')

const log = debug('@modernpoacher/gremlins/gremlins/textbox')

log('`gremlins` is awake')

const {
  default: component
} = require('./index.jsx')

module.exports = component
