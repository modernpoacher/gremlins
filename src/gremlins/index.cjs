require('@babel/register')({
  ignore: [
    /node_modules\/(?!@modernpoacher\/cogs|@modernpoacher\/gremlins|@modernpoacher\/sprockets)/
  ]
})

const debug = require('debug')

const log = debug('@modernpoacher/gremlins/gremlins')

log('`gremlins` is awake')

/**
 *  Has default export, etc
 */
module.exports = require('./index.jsx')
