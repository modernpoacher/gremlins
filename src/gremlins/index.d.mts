declare module '#gremlins/gremlins' {
  import React from 'react'

  export interface FieldGremlinProps {
    id: string
    name: string
    required: boolean
    disabled: boolean
    readOnly: boolean
    placeholder: string
    onChange: GremlinsTypes.OnChangeType
    fieldRef: object
  }

  export interface ValueGremlinProps extends FieldGremlinProps {
    value: string
    defaultValue: string
    onChange: GremlinsTypes.OnChangeType
  }

  export interface CheckGremlinProps extends FieldGremlinProps {
    checked: boolean
    defaultChecked: boolean
    onClick: GremlinsTypes.OnClickType
  }

  export interface GroupGremlinProps {
    title: string
    onChange: GremlinsTypes.OnChangeType
    groupRef: object
    children: React.JSX.Element | React.JSX.Element[]
  }

  export class FieldGremlin extends React.Component<FieldGremlinProps> {}

  export class ValueGremlin extends React.Component<ValueGremlinProps> {}

  export class CheckGremlin extends React.Component<CheckGremlinProps> {}

  export class GroupGremlin extends React.Component<GroupGremlinProps> {}
}

declare module '@modernpoacher/gremlins/gremlins' {
  export * from '#gremlins/gremlins'
}
