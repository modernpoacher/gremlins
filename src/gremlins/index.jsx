/**
 * GroupGremlin component
 *
 * @typedef {import('@modernpoacher/gremlins/gremlins').GroupGremlinProps} GroupGremlinProps
 */
/**
 * FieldGremlin component
 *
 * @typedef {import('@modernpoacher/gremlins/gremlins').FieldGremlinProps} FieldGremlinProps
 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import Field from '@modernpoacher/gremlins/components/field'
import Group from '@modernpoacher/gremlins/components/group'

function onChange () {
  /* */
}

function onClick () {
  /* */
}

export class FieldGremlin extends Component {
  getClassName () {
    const {
      required,
      disabled,
      readOnly
    } = this.props

    return classnames('gremlin', { required, disabled, readOnly })
  }

  getId () {
    const {
      id,
      name
    } = this.props

    return id || name
  }

  /**
   * @param {FieldGremlinProps} props
   * @returns {boolean}
   */
  shouldComponentUpdate (props) {
    return (
      (props.name !== this.props.name) ||
      (props.id !== this.props.id) ||
      (props.required !== this.props.required) ||
      (props.disabled !== this.props.disabled) ||
      (props.readOnly !== this.props.readOnly) ||
      (props.tabIndex !== this.props.tabIndex) ||
      (props.accessKey !== this.props.accessKey) ||
      (props.placeholder !== this.props.placeholder) ||
      (props.onChange !== this.props.onChange)
    )
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder,
      onChange,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={onChange}
        fieldRef={fieldRef}
      />
    )
  }

  render () {
    const className = this.getClassName()

    return (
      <div className={className}>
        {this.renderField()}
      </div>
    )
  }
}

FieldGremlin.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  readOnly: PropTypes.bool,
  placeholder: PropTypes.string,
  onChange: PropTypes.func,
  fieldRef: PropTypes.shape({
    current: PropTypes.shape().isRequired
  })
}

FieldGremlin.defaultProps = {
  required: false,
  disabled: false,
  readOnly: false,
  onChange
}

export class ValueGremlin extends FieldGremlin {
  shouldComponentUpdate (props) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.value !== this.props.value)
    )
  }
}

ValueGremlin.propTypes = {
  ...FieldGremlin.propTypes,
  value: PropTypes.string,
  defaultValue: PropTypes.string
}

ValueGremlin.defaultProps = {
  ...FieldGremlin.defaultProps
}

export class CheckGremlin extends FieldGremlin {
  shouldComponentUpdate (props) {
    return (
      super.shouldComponentUpdate(props) ||
      (props.checked !== this.props.checked) ||
      (props.onClick !== this.props.onClick)
    )
  }
}

CheckGremlin.propTypes = {
  ...FieldGremlin.propTypes,
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  onClick: PropTypes.func
}

CheckGremlin.defaultProps = {
  ...FieldGremlin.defaultProps,
  onClick
}

export class GroupGremlin extends Component {
  getClassName () {
    return 'gremlin'
  }

  /**
   * @param {GroupGremlinProps} props
   * @returns {boolean}
   */
  shouldComponentUpdate (props) {
    return (
      (props.children !== this.props.children) ||
      (props.onChange !== this.props.onChange)
    )
  }

  renderGroup () {
    const {
      onChange,
      groupRef,
      children
    } = this.props

    return (
      <Group
        onChange={onChange}
        groupRef={groupRef}>
        {children}
      </Group>
    )
  }

  render () {
    const className = this.getClassName()

    return (
      <div className={className}>
        {this.renderGroup()}
      </div>
    )
  }
}

GroupGremlin.propTypes = {
  onChange: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(
      PropTypes.node
    )
  ]),
  groupRef: PropTypes.shape({
    current: PropTypes.shape().isRequired
  })
}

GroupGremlin.defaultProps = {
  onChange
}
