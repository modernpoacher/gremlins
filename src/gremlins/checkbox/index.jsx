/**
 * CheckboxGremlin component
 */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { CheckGremlin } from '@modernpoacher/gremlins/gremlins'

import Field from './field/index.jsx'

export default class CheckboxGremlin extends CheckGremlin {
  getClassName () {
    return classnames(super.getClassName(), 'checkbox')
  }

  handleClick = (value) => {
    const {
      onClick
    } = this.props

    onClick(value)
  }

  handleChange = (value) => {
    const {
      onChange
    } = this.props

    onChange(value)
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      value,
      checked,
      defaultChecked,
      required = false,
      disabled = false,
      readOnly = false,
      tabIndex,
      accessKey,
      placeholder,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        value={value}
        checked={checked}
        defaultChecked={defaultChecked}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onClick={this.handleClick}
        onChange={this.handleChange}
        fieldRef={fieldRef}
      />
    )
  }

  render () {
    const className = this.getClassName()

    return (
      <div className={className}>
        {this.renderField()}
      </div>
    )
  }
}

CheckboxGremlin.propTypes = {
  ...CheckGremlin.propTypes,
  value: PropTypes.string.isRequired
}

CheckboxGremlin.defaultProps = {
  ...CheckGremlin.defaultProps
}
