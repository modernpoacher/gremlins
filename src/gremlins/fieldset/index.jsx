/**
 * FieldsetGremlin component
 *
 * @typedef {import('@modernpoacher/gremlins/gremlins').GroupGremlinProps} GroupGremlinProps
 */

import React from 'react'

import classnames from 'classnames'

import { GroupGremlin as Gremlin } from '@modernpoacher/gremlins/gremlins'
import Group from './group/index.jsx'

export default class FieldsetGremlin extends Gremlin {
  getClassName () {
    return classnames(super.getClassName(), 'fieldset')
  }

  renderGroup () {
    const {
      onChange,
      children
    } = this.props

    return (
      <Group
        onChange={onChange}>
        {children}
      </Group>
    )
  }
}

FieldsetGremlin.propTypes = {
  ...Gremlin.propTypes
}

FieldsetGremlin.defaultProps = {
  ...Gremlin.defaultProps
}
