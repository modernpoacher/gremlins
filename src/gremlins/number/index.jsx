/**
 * NumberGremlin component
 */
import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { ValueGremlin } from '@modernpoacher/gremlins/gremlins'

import Field from './field/index.jsx'

export default class NumberGremlin extends ValueGremlin {
  getClassName () {
    return classnames(super.getClassName(), 'number')
  }

  handleChange = (value) => {
    const {
      onChange
    } = this.props

    onChange(value)
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        fieldRef={fieldRef}
      />
    )
  }
}

NumberGremlin.propTypes = {
  ...ValueGremlin.propTypes,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  defaultValue: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ])
}

NumberGremlin.defaultProps = {
  ...ValueGremlin.defaultProps
}
