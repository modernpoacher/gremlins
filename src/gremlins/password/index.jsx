/**
 * PasswordGremlin component
 */
import React from 'react'
import classnames from 'classnames'

import { ValueGremlin } from '@modernpoacher/gremlins/gremlins'

import Field from './field/index.jsx'

export default class PasswordGremlin extends ValueGremlin {
  getClassName () {
    return classnames(super.getClassName(), 'password')
  }

  handleChange = (value) => {
    const {
      onChange
    } = this.props

    onChange(value)
  }

  renderField () {
    const id = this.getId()

    const {
      name,
      value,
      defaultValue,
      required,
      disabled,
      readOnly,
      tabIndex,
      accessKey,
      placeholder,
      fieldRef
    } = this.props

    return (
      <Field
        id={id}
        name={name}
        value={value}
        defaultValue={defaultValue}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        tabIndex={tabIndex}
        accessKey={accessKey}
        placeholder={placeholder}
        onChange={this.handleChange}
        fieldRef={fieldRef}
      />
    )
  }
}

PasswordGremlin.propTypes = {
  ...ValueGremlin.propTypes
}

PasswordGremlin.defaultProps = {
  ...ValueGremlin.defaultProps
}
