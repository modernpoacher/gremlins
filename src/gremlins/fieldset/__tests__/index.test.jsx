import React from 'react'
import renderer from 'react-test-renderer'

import classnames from 'classnames'

import { GroupGremlin as Super } from '@modernpoacher/gremlins/gremlins'
import Gremlin from '../index.jsx'

jest.mock('classnames', () => jest.fn(() => 'MOCK CLASSNAME'))

jest.mock('../group/index.jsx')

describe('@modernpoacher/gremlins/gremlins/fieldset', () => {
  describe('<Gremlin />', () => {
    describe('With required props', () => {
      const component = (
        <Gremlin />
      )

      it('renders', () => {
        return expect(renderer.create(component).toJSON())
          .toMatchSnapshot()
      })

      describe('`getClassName`', () => {
        it('is defined', () => {
          return expect(Gremlin.prototype.getClassName)
            .toBeDefined()
        })
      })

      describe('`renderGroup`', () => {
        it('is defined', () => {
          return expect(Gremlin.prototype.renderGroup)
            .toBeDefined()
        })
      })
    })

    describe('With additional props', () => {
      it('renders', () => {
        const component = (
          <Gremlin
            onChange={jest.fn()}
          />
        )

        return expect(renderer.create(component).toJSON())
          .toMatchSnapshot()
      })
    })

    describe('`getClassName()`', () => {
      let returnValue

      beforeEach(() => {
        jest.spyOn(Super.prototype, 'getClassName').mockReturnValue('MOCK GETCLASSNAME')

        const component = (
          <Gremlin />
        )

        const instance = (
          renderer.create(component)
            .getInstance()
        )

        returnValue = instance.getClassName()
      })

      it('invokes `classnames`', () => {
        return expect(classnames)
          .toBeCalledWith('MOCK GETCLASSNAME', 'fieldset')
      })

      it('returns the classname', () => {
        return expect(returnValue)
          .toBe('MOCK CLASSNAME')
      })
    })
  })
})
