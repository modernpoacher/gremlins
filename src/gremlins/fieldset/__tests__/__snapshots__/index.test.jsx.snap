// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modernpoacher/gremlins/gremlins/fieldset <Gremlin /> With additional props renders 1`] = `
<div
  className="MOCK CLASSNAME"
/>
`;

exports[`@modernpoacher/gremlins/gremlins/fieldset <Gremlin /> With required props renders 1`] = `
<div
  className="MOCK CLASSNAME"
/>
`;
